### PR TITLE
[Feature] Support multi-column IN subqueries in WHERE clause

### PR DESCRIPTION
**Pull request by agubichev**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes # https://github.com/StarRocks/starrocks/issues/14404

## Problem Summary(Required) ：
This adds a multi-column IN subquery support into the optimizer. It uses a separate MultiInPredicate to represent such a subquery, and rewrites it into a semi- (anti-) join.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@agubichev Hi, you need to config the code style according to `fe/checkstyle.xml`, we have a force check of code style in UT

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+package com.starrocks.analysis;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java:1]: Please add the license header

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java [hunk 2]
```diff
@@ -51,27 +53,51 @@ public boolean check(OptExpression input, OptimizerContext context) {
                 && !SubqueryUtils.containsCorrelationSubquery(input);
     }
 
+    private static ScalarOperator normalizeBinaryOperator(BinaryPredicateOperator input) {
+        if (input.getUsedColumns().isEmpty()) {
+            return new FoldConstantsRule().visitBinaryPredicate(input, null);
+        } else if (input.getChild(0).getUsedColumns().isEmpty()) {
+            // normalize binaryPredicateOperator, e.g. 1 = tbl.v1 to tbl.v1 = 1
+            input.swap();
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:61]: You can do this with NormalizePredicateRule

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java [hunk 3]
```diff
@@ -51,27 +53,51 @@ public boolean check(OptExpression input, OptimizerContext context) {
...
+            List<ScalarOperator> conjuncts = Lists.newArrayList();
+            ScalarOperatorRewriter rewriter = new ScalarOperatorRewriter();
+            for (int i = 0; i < multiIn.getTupleSize(); ++i) {
+                ScalarOperator left = multiIn.getChild(i);
+                ScalarOperator right = multiIn.getChild(multiIn.getTupleSize() + i);
+                ScalarOperator conjunct =
+                        normalizeBinaryOperator(new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.EQ,
+                                left, right));
+                ScalarOperator conjunctWithCast =
+                        rewriter.rewrite(conjunct, ScalarOperatorRewriter.DEFAULT_TYPE_CAST_RULE);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:83]: do cast/normalize/foldconstant with scalar operator rewrite rule

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ * Represents an IN subquery with multiple columns.
+ */
+public class MultiInPredicate extends Predicate{
+    private final boolean isNotIn;
+    // Expressions involved on the left side of the IN predicate (outside the subquery).
+    List<Expr> outerExprs;
+    Expr subquery;
+
+    // Number of columns involved in the IN subquery.
+    int tupleSize;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java:19]: private

Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java:19]: `columnSize` is better?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java (hunk 3 above)
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:83]: do cast with implicitCastRule plz, cast affects correctness, it's best to deal with it early

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 5]
```diff
@@ -1672,6 +1672,11 @@ booleanExpression
 
 predicate
     : valueExpression (predicateOperations[$valueExpression.ctx])?
+    | tupleInSubquery
+    ;
+
+tupleInSubquery
+    : '(' (expression (',' expression)*)? ')' NOT? IN '(' queryRelation ')'
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1679]: should be?
```suggestion
    : '(' expression (',' expression)+ ')' NOT? IN '(' queryRelation ')'
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java [hunk 6]
```diff
@@ -51,27 +53,51 @@ public boolean check(OptExpression input, OptimizerContext context) {
...
+                conjuncts.add(conjunctWithCast);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:84]: do cast before fold constant

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:61]: subquery maybe can't do that now

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java [hunk 7]
```diff
@@ -528,6 +530,58 @@ public ScalarOperator visitTimestampArithmeticExpr(TimestampArithmeticExpr node,
...
+            ((Subquery) node.getChild(node.getTupleSize())).setUseSemiAnti(context.useSemiAnti);
+
+            LogicalApplyOperator applyOperator = LogicalApplyOperator.builder().setOutput(outputPredicateRef)
+                    .setSubqueryOperator(inPredicateOperator)
+                    .setCorrelationColumnRefs(subqueryPlan.getCorrelation())
+                    .setUseSemiAnti(context.useSemiAnti).build();
+
+            SubqueryOperator subqueryOperator = new SubqueryOperator(rightColRefs.get(0).getType(), queryStatement,
+                    applyOperator, subqueryPlan.getRootBuilder());
+
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:580]: here is to deceive the optimizer, the type of subquery can't be the first column’s type in actually, and it's need don't handle MultiInOperator cast in ImplicitCastRule, I think it's a trick and will cause other problems... you need more test and comment it

File: fe/fe-core/src/test/resources/sql/subquery/in-subquery.sql [hunk 8]
```diff
@@ -886,3 +886,101 @@ NULL AWARE LEFT ANTI JOIN (join-predicate [19: cast = 20: abs AND add(cast(9: t1
         SCAN (columns[7: t1a, 9: t1c, 10: t1d] predicate[null])
 [end]
 
+[sql]
```
Comment (Seaven) [inline fe/fe-core/src/test/resources/sql/subquery/in-subquery.sql:889]: add test to cover (int, String) in (string, double)....more implict cast case

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java:61]: you can do this with ScalarOperatorRewriter's DEFAULT_REWRITE_RULES

**Comment by chaoyli** [comment]
how to implement a query like `select * from tableA where (col1, col2) in (select 1, 2)`;

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 7 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:583]: should i add a special TupleType?
I agree it's a bit of a trick, but this subquery operator should be fully removed by one of the "earlier" logical rules (like QuantifiedApply2JoinRule). We could add a check that the MultiInPredicateOperator is not in the query plan after those subquery rewriting rules.
I modified the ImplicitCastRule to throw an exception if the MultiInPredicate is encountered.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java (hunk 1 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java:15]: done

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java (hunk 4 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java]: renamed to numberOfColumns.

File: fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            if (i < tupleSize - 1) {
+                strBuilder.append(", ");
+            }
+        }
+        strBuilder.append(") ").append(notStr).append("IN (").append(getChild(tupleSize).toSql());
+        strBuilder.append(")");
+        return strBuilder.toString();
+    }
+
+    public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/analysis/MultiInPredicate.java:80]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 10]
```diff
@@ -625,6 +626,32 @@ public Void visitInPredicate(InPredicate node, Scope scope) {
...
+
+            Subquery inSubquery = (Subquery) node.getChild(node.getTupleSize());
+            List<Type> rightTypes =
+                    inSubquery.getQueryStatement().getQueryRelation().getOutputExpression().stream().map(Expr::getType).collect(Collectors.toList());
+            if (leftTypes.size() != rightTypes.size()) {
+                throw new SemanticException("subquery must return the same number of columns as provided by the IN predicate");
+            }
+
+            for (int i = 0; i < rightTypes.size(); ++i) {
+                if (leftTypes.get(i).isJsonType() || rightTypes.get(i).isJsonType()) {
```
[the remark this answers is not in the document]
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java (hunk 2 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java]: done, replaced with DEFAULT_REWRITE_RULES

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 5 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: done

File: fe/fe-core/src/test/resources/sql/subquery/in-subquery.sql (hunk 8 above)
Comment (agubichev) [reply fe/fe-core/src/test/resources/sql/subquery/in-subquery.sql:889]: done

**Comment by chaoyli** [comment]
@agubichev

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 7 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:583]: I think don't need add speical TupleType now, but you'd better add comment here. btw, it's a good idea to throw an exception in ImplicitCastRule

**Comment by agubichev** [comment]
> @agubichev If use the left semi join to do this, it may be a little non efficient? image

It is not as efficient as having a Tuple predicate over customer, but it's better than "unimplemented error". We can consider rewriting it to a tuple predicate over table scan once those are supported by the execution system.

**Comment by agubichev** [comment]
> run starrocks_admit_test

Where do i find it?

**Comment by agubichev** [comment]
> **starrocks_admit_test ** Expected — Waiting for status to be reported

looks like that workflow needs to be approved?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 7 above)
Comment (agubichev) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:583]: comment regarding the type is added

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by chaoyli** [comment]
> > @agubichev If use the left semi join to do this, it may be a little non efficient? image
> 
> It is not as efficient as having a Tuple predicate over customer, but it's better than "unimplemented error". We can consider rewriting it to a tuple predicate over table scan once those are supported by the execution system.

@agubichev I get it. So the `select * from tableA where (col1, col2) in (select 1, 2 union select 3, 4);` is the same?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 112 / 153 (73.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|0|1|00.00%|[188]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/MultiInPredicateOperator.java|9|29|31.03%|[50, 51, 52, 53, 54, 55, 58, 59, 60, 61, 66, 71, 72, 74, 75, 77, 78, 80, 81, 86]|
|:large_blue_circle:|com/starrocks/analysis/MultiInPredicate.java|22|37|59.46%|[47, 48, 49, 50, 54, 81, 82, 84, 85, 87, 88, 89, 91, 96, 101]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|16|18|88.89%|[650, 651]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|32|35|91.43%|[549, 560, 566]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java|23|23|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|1|1|100.00%|[]|

**Comment by agubichev** [comment]
> @agubichev I get it. So the `select * from tableA where (col1, col2) in (select 1, 2 union select 3, 4);` is the same?

Correct, in both cases the best plan would be something like

Filter( TupleFilter(col1, col2, ((1,2), (3,4)) ))
|
|
Scan(TableA)

**Review by Seaven — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
The sonar complains about `MultiInPredicate` overwrites `equals` but not the `hashCode`, could you please fix them by the way

**Comment by agubichev** [comment]
> The sonar complains about `MultiInPredicate` overwrites `equals` but not the `hashCode`, could you please fix them by the way

Done

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by chaoyli** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 111 / 152 (73.03%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java|0|1|00.00%|[188]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/MultiInPredicateOperator.java|9|29|31.03%|[50, 51, 52, 53, 54, 55, 58, 59, 60, 61, 66, 71, 72, 74, 75, 77, 78, 80, 81, 86]|
|:large_blue_circle:|com/starrocks/analysis/MultiInPredicate.java|23|38|60.53%|[48, 49, 50, 51, 55, 82, 83, 85, 86, 88, 89, 90, 92, 102, 107]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|15|17|88.24%|[636, 637]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|32|35|91.43%|[549, 560, 566]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2JoinRule.java|23|23|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/QuantifiedApply2OuterJoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|1|1|100.00%|[]|

**Pull request merged**